storage/pipeline/piece/piece_info.go

```diff
@@ -0,0 +1,186 @@
+}
+
+// Valid validates the deal info after being accepted through RPC, checks that
+// the deal metadata is well-formed.
+func (ds *PieceDealInfo) Valid(nv network.Version) error {
+	hasLegacyDealInfo := ds.PublishCid != nil && ds.DealID != 0 && ds.DealProposal != nil
+	hasPieceActivationManifest := ds.PieceActivationManifest != nil
+
+	if hasLegacyDealInfo && hasPieceActivationManifest {
+		return xerrors.Errorf("piece deal info has both legacy deal info and piece activation manifest")
+	}
+
+	if !hasLegacyDealInfo && !hasPieceActivationManifest {
+		return xerrors.Errorf("piece deal info has neither legacy deal info nor piece activation manifest")
+	}
+
+	if hasLegacyDealInfo {
+		if _, err := ds.DealProposal.Cid(); err != nil {
+			return xerrors.Errorf("checking proposal CID: %w", err)
+		}
+	}
+
+	if ds.DealSchedule.StartEpoch <= 0 {
+		return xerrors.Errorf("invalid deal start epoch %d", ds.DealSchedule.StartEpoch)
+	}
+	if ds.DealSchedule.EndEpoch <= 0 {
+		return xerrors.Errorf("invalid deal end epoch %d", ds.DealSchedule.EndEpoch)
+	}
+	if ds.DealSchedule.EndEpoch <= ds.DealSchedule.StartEpoch {
+		return xerrors.Errorf("invalid deal end epoch %d (start %d)", ds.DealSchedule.EndEpoch, ds.DealSchedule.StartEpoch)
+	}
+
+	if hasPieceActivationManifest {
+		if nv < network.Version22 {
+			return xerrors.Errorf("direct-data-onboarding pieces aren't accepted before network version 22")
+		}
+
+		// todo any more checks seem reasonable to put here?
+	}
+
+	return nil
+}
+
+type AllocationAPI interface {
+	StateGetAllocationForPendingDeal(ctx context.Context, dealId abi.DealID, tsk types.TipSetKey) (*verifregtypes.Allocation, error)
+	StateGetAllocation(ctx context.Context, clientAddr address.Address, allocationId verifregtypes.AllocationId, tsk types.TipSetKey) (*verifregtypes.Allocation, error)
+}
+
+func (ds *PieceDealInfo) GetAllocation(ctx context.Context, aapi AllocationAPI, tsk types.TipSetKey) (*verifregtypes.Allocation, error) {
+	switch {
+	case ds.isBuiltinMarketDeal():
+		return aapi.StateGetAllocationForPendingDeal(ctx, ds.DealID, tsk)
+	default:
+		if ds.PieceActivationManifest.VerifiedAllocationKey == nil {
+			return nil, nil
+		}
+
+		caddr, err := address.NewIDAddress(uint64(ds.PieceActivationManifest.VerifiedAllocationKey.Client))
+		if err != nil {
+			return nil, err
+		}
+
+		all, err := aapi.StateGetAllocation(ctx, caddr, verifregtypes.AllocationId(ds.PieceActivationManifest.VerifiedAllocationKey.ID), tsk)
+		if err != nil {
+			return nil, err
+		}
+
+		if all == nil {
+			return nil, nil
+		}
+
+		if all.Client != ds.PieceActivationManifest.VerifiedAllocationKey.Client {
+			return nil, xerrors.Errorf("allocation client mismatch: %d != %d", all.Client, ds.PieceActivationManifest.VerifiedAllocationKey.Client)
+		}
+
+		return all, nil
+	}
+}
+
+// StartEpoch returns the last epoch in which the sector containing this deal
+// must be sealed in order for the deal to be valid. (todo review: precommitted or committed??)
```

Comment:
Pretty sure this is Committed, precommit doesn't matter in the ddo world

Reply:
(that's because precommit will now only say 'I claim I'm able to store this bit of data' without referencing anything market-related, and commit is the one that tells the market actor that data is now stored in a sealed sector)